controllers/helmrepository_controller_oci.go

```diff
@@ -0,0 +1,362 @@
+		if err := r.Client.Get(ctx, name, &secret); err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to get secret '%s': %w", name.String(), err),
+				Reason: sourcev1.AuthenticationFailedReason,
+			}
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+
+		// Construct actual options
+		logOpt, err := loginOptionFromSecret(secret)
+		if err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to configure Helm client with secret data: %w", err),
+				Reason: sourcev1.AuthenticationFailedReason,
+			}
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			// Return err as the content of the secret may change.
+			return sreconcile.ResultEmpty, e
+		}
+
+		logOpts = append(logOpts, logOpt)
+	}
+
+	if result, err := r.validateSource(ctx, obj, logOpts...); err != nil || result == sreconcile.ResultEmpty {
+		return result, err
+	}
+
+	return sreconcile.ResultSuccess, nil
+}
+
+// validateSource the HelmRepository object by checking the url and connecting to the underlying registry
+// with he provided credentials.
+func (r *HelmRepositoryOCIReconciler) validateSource(ctx context.Context, obj *sourcev1.HelmRepository, logOpts ...registry.LoginOption) (sreconcile.Result, error) {
+	registryClient, file, err := r.RegistryClientGenerator(logOpts != nil)
+	if err != nil {
+		e := &serror.Stalling{
+			Err:    fmt.Errorf("failed to create registry client:: %w", err),
+			Reason: meta.FailedReason,
+		}
+		conditions.MarkFalse(obj, meta.ReadyCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+
+	if file != "" {
+		defer func() {
+			os.Remove(file)
+		}()
+	}
+
+	chartRepo, err := repository.NewOCIChartRepository(obj.Spec.URL, repository.WithOCIRegistryClient(registryClient))
+	if err != nil {
+		if strings.Contains(err.Error(), "parse") {
+			e := &serror.Stalling{
+				Err:    fmt.Errorf("failed to parse URL '%s': %w", obj.Spec.URL, err),
+				Reason: sourcev1.URLInvalidReason,
+			}
+			conditions.MarkFalse(obj, meta.ReadyCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		} else if strings.Contains(err.Error(), "the url scheme is not supported") {
+			e := &serror.Event{
+				Err:    err,
+				Reason: sourcev1.URLInvalidReason,
+			}
+			conditions.MarkFalse(obj, meta.ReadyCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+	}
+
+	// Attempt to login to the registry if credentials are provided.
+	if logOpts != nil {
+		err = chartRepo.Login(logOpts...)
+		if err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to create temporary file: %w", err),
+				Reason: meta.FailedReason,
+			}
+			conditions.MarkFalse(obj, meta.ReadyCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+	}
+
+	conditions.MarkTrue(obj, meta.ReadyCondition, meta.SucceededReason, "Helm repository %q is ready", obj.Name)
```

Comment:
Thinking more about how to make Ready condition True... since we don't have any positive polarity condition to depend on, we may have to directly set Ready as True. I thought about introducing a positive condition like `RegistryConnected` or `RegistryAccessible`, but it didn't feel much different from a `FetchFailed` condition, fetch happens only after logging in or when it's accessible, so it felt redundant.

I think it'd be better to set Ready condition based on the FetchFailed condition. So, maybe we can add a defer function at the beginning of this function, `validateSource()`,  in which we check if FetchFailed is not True or unset, and then set Ready to True. Which means if any of the checks above sets FetchFailed to True, Ready is not touched. If none of the checks fail and any existing FetchFailed condition is removed, we set Ready to True.
Ready will be automatically set to False in summarize and patch if negative polarity condition, FetchFailed, is True.

Reply:
On the first iteration we had a specific conditions for this validation loop, but after the reviews we decided that using only the `readyCondition` was acceptable. I stated above we do not fetch anything so using `FetchFailed` seems odd.